matrix-sdk-android/src/main/java/org/matrix/android/sdk/internal/crypto/MXOlmDevice.kt

```diff
@@ -429,33 +428,34 @@ internal class MXOlmDevice @Inject constructor(
     }
 
     fun storeOutboundGroupSessionForRoom(roomId: String, sessionId: String) {
-        outboundGroupSessionStore[sessionId]?.let {
+        outboundGroupSessionCache[sessionId]?.let {
             store.storeCurrentOutboundGroupSessionForRoom(roomId, it)
         }
     }
 
-    fun restoreOutboundGroupSessionForRoom(roomId: String):  MXOutboundSessionInfo? {
+    fun restoreOutboundGroupSessionForRoom(roomId: String): MXOutboundSessionInfo? {
         val restoredOutboundGroupSession = store.getCurrentOutboundGroupSessionForRoom(roomId)
         if (restoredOutboundGroupSession != null) {
             val sessionId = restoredOutboundGroupSession.outboundGroupSession.sessionIdentifier()
-            if (!outboundGroupSessionStore.containsKey(sessionId)) {
-                outboundGroupSessionStore[sessionId] = restoredOutboundGroupSession.outboundGroupSession
-                return MXOutboundSessionInfo(
-                        sessionId = sessionId,
-                        sharedWithHelper = SharedWithHelper(roomId, sessionId, store),
-                        restoredOutboundGroupSession.creationTime
-                )
-            } else {
-                restoredOutboundGroupSession.outboundGroupSession.releaseSession()
-            }
+            // cache it
+            outboundGroupSessionCache[sessionId] = restoredOutboundGroupSession.outboundGroupSession
+
+            return MXOutboundSessionInfo(
+                    sessionId = sessionId,
+                    sharedWithHelper = SharedWithHelper(roomId, sessionId, store),
+                    restoredOutboundGroupSession.creationTime
+            )
         }
         return null
     }
 
     fun discardOutboundGroupSessionForRoom(roomId: String) {
-        outboundGroupSessionStore.remove(roomId)?.releaseSession()
+        store.getCurrentOutboundGroupSessionForRoom(roomId)?.outboundGroupSession?.sessionIdentifier()?.let { sessionId ->
+            outboundGroupSessionCache.remove(sessionId)?.releaseSession()
```

Comment:
Do you have to also release what have been returned by `store.getCurrentOutboundGroupSessionForRoom(roomId)`?